homeassistant/components/esphome/config_flow.py

```diff
@@ -89,23 +90,34 @@ async def async_step_zeroconf(self, user_input: ConfigType):
         address = user_input["properties"].get("address", local_name)
 
         # Check if already configured
+        await self.async_set_unique_id(node_name)
+        self._abort_if_unique_id_configured(updates={CONF_HOST: address})
+
         for entry in self._async_current_entries():
             already_configured = False
-            if entry.data["host"] == address:
+
+            if entry.data[CONF_HOST] == address:
                 # Is this address already configured?
                 already_configured = True
             elif entry.entry_id in self.hass.data.get(DATA_KEY, {}):
                 # Does a config entry with this name already exist?
                 data: RuntimeEntryData = self.hass.data[DATA_KEY][entry.entry_id]
+
                 # Node names are unique in the network
                 if data.device_info is not None:
                     already_configured = data.device_info.name == node_name
 
             if already_configured:
+                # Backwards compat, we update old entries
+                if not entry.unique_id:
+                    self.hass.config_entries.async_update_entry(
+                        entry, unique_id=node_name
+                    )
```

Comment:
We should move this part into the entry setup. The entry is already aware of this and can update the entry directly. That way, we don't have to rely on discovery for updating it. Furthermore, we can remove all the above code (including address matching and searching).

Reply:
Second thought, this place is perfect, as the config entry does not have this value and things might differ.

One suggestion for improving the hit/miss factor here:

Maybe we should try resolving the given address to an IP and see if that IP address matches. This would cover the case where people have manually entered an IP address before.

Reply:
i believe that second part would cover the IP being used since it queries the device info?

https://github.com/home-assistant/core/pull/34753/files#diff-eab03bb3885f0680387552cf5fc738bbR102

Reply:
that said i do wonder if the config entry should store the discovery resolved host rather than the .local as its host? we could have that slowly fix itself via the abort update now that unique id would be set.

Reply:
made some updates and if address isn't entry host then we check if discovered IP address is since that would catch manual configs. if the IP doesnt match either. the other clause that looks in device_info seems to catch updated IPs from the test `test_discovery_already_configured_name`